feat: kapt.include.compile.classpath for Maven (#5575)
JetBrains/kotlin-web-site: docs/topics/compiler-plugins/kapt.md

[INFO] org.mapstruct.ap.MappingProcessor: total sources: 2, sources per round: 2, 0, 0
```

### Exclude annotation processors from compile classpath

> You can disable the discovery of annotation processors that are not included in the kapt's processor path.

Reviewer: ```suggestion
You can disable the discovery of annotation processors that aren't included in kapt's processor path.
```

This effectively excludes unnecessary annotation processors from the compile classpath.

#### In Gradle

Gradle uses [compile avoidance](https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_compile_avoidance)
to skip annotation processing during project rebuild, improving incremental build times with kapt. Particularly,
annotation processing is skipped when:

* The project's source files are unchanged.
* The changes in dependencies are [ABI](https://en.wikipedia.org/wiki/Application_binary_interface)-compatible.
  For example, the only changes are in method bodies.

However, compile avoidance can't be used for annotation processors discovered on the compile classpath, since changes
in their internal implementation require running the annotation processing tasks, even if the ABI remains unchanged.

> That's why it's not recommended to use annotation processors from the compile classpath. To exclude such annotations

Reviewer: ```suggestion
That's why we don't recommend using annotation processors from the compile classpath. To exclude these annotations
```
We don't recommend, right?

Author: Well, 'we' is used rarely outside WN and promotional material

from processing, add the `kapt.include.compile.classpath` property to your `gradle.properties` file:

```none
# gradle.properties
kapt.include.compile.classpath=false
```

> With the option set to `false`, annotation processor dependencies that are not included in the processor path

Reviewer: ```suggestion
With the option set to `false`, annotation processor dependencies that aren't included in the processor path
```

> (the `kapt*` configurations) are excluded from the kapt processing.

Reviewer: ```suggestion
(the `kapt*` configurations) are excluded from kapt processing.
```

#### In Maven

> To exclude annotation processors that are missing from the kapt's processor path, set the `includeCompileClasspath`

Reviewer: ```suggestion
To exclude annotation processors that are missing from kapt's processor path, set the `includeCompileClasspath`
```

option to `false` in the `<execution>` section of the kapt plugin:

```xml
<execution>
    <id>kapt</id>
    <goals>
        <goal>kapt</goal>
    </goals>
    <configuration>
        <includeCompileClasspath>false</includeCompileClasspath>
        <sourceDirs>...</sourceDirs>
        <annotationProcessorPaths>...</annotationProcessorPaths>
    </configuration>
</execution>
```

Alternatively, you can use the `kapt.include.compile.classpath` property in the `<properties>` section of your `pom.xml`:

```xml
<properties>
    <kapt.include.compile.classpath>false</kapt.include.compile.classpath>
</properties>
```

> With the option set to `false`, annotation processors not included in the `<annotationProcessorPaths>` section are

Reviewer: ```suggestion
With the option set to `false`, annotation processors that aren't included in the `<annotationProcessorPaths>` section are
```

excluded from kapt processing.

> If `includeCompileClasspath` is not set and kapt detects an annotation processor on the compile classpath that is not

Reviewer: ```suggestion
If the `includeCompileClasspath` option isn't set and kapt detects an annotation processor on the compile classpath that isn't
```

explicitly defined in the processor path, you'll see a deprecation warning:

```none
[WARNING] Annotation processors discovery from compile classpath is deprecated.
Set 'kapt.include.compile.classpath=false' to disable discovery.
```

> > To see the list of annotation processors that are not present on the kapt classpath, run the build with the `--info` log

Reviewer: ```suggestion
> To see the list of annotation processors that aren't present on the kapt classpath, run the build with the `--info` log
```

[INFO] org.mapstruct.ap.MappingProcessor: total sources: 2, sources per round: 2, 0, 0
```

### Exclude annotation processors from compile classpath

You can disable the discovery of annotation processors that are not included in the kapt's processor path.
This effectively excludes unnecessary annotation processors from the compile classpath.

#### In Gradle

> In Gradle, excluding annotation processors from the compile classpath enables Gradle's [compile avoidance](https://docs.gradle.org/current/userguide/java_plugin.html#sec:java_compile_avoidance).

Reviewer 2: I am not quite sure that this sentence is correct. In my understanding, the Gradle's compile avoidance is enabled by default - it is just a way how Gradle's plugin works (unless `clean` target is explicitly specified). And due to that fact, it is not recommended to use annotations processors from the compile classpath, as changes in their internal implementation may be not applied to other modules if ABI is not changed.

Author: Thank you for the clarification! rephrased the Gradle part
